pensieve/statistics.py

```diff
@@ -0,0 +1,160 @@
+
+
+@attr.s(auto_attribs=True)
+class StatisticResult:
+    """
+    Represents the resulting data after applying a statistic transformation
+    to metric data.
+    """
+
+    metric: str
+    statistic: str
+    parameter: float
+    label: str
+    ci_width: Optional[float] = 0.0
+    point: Optional[float] = 0.0
+    lower: Optional[float] = 0.0
+    upper: Optional[float] = 0.0
+
+    def with_ci(self, data: DataFrame, t: float, confidence_level: float) -> "StatisticResult":
+        """Calculate the confidence interval and update result."""
+        confidence_margin = 0.5 * (1.0 - confidence_level)
+        confidence_high = (0.0 + confidence_margin) * 100
+        confidence_low = (1.0 - confidence_margin) * 100
+        self.lower = t - np.percentile(data - t, confidence_low)
+        self.upper = t - np.percentile(data - t, confidence_high)
+        self.ci_width = confidence_level
+        self.point = t
+        return self
+
+    def with_point(self, point_value: float) -> "StatisticResult":
+        """Set provided value as point value result for statistic."""
+        self.point = point_value
+        return self
+
+
+@attr.s(auto_attribs=True)
+class StatisticResultCollection:
+    """
+    Represents a set of statistics result data.
+    """
+
+    data: List[StatisticResult] = []
+
+    def append(self, result: StatisticResult):
+        self.data.append(result)
+
+    def merge(self, result_collection: "StatisticResultCollection"):
+        self.data = self.data + result_collection.data
+
+    def save_to_bigquery(self, client, destination_table, append=True):
+        """Stores the data to a BigQuery table with a defined schema."""
+
+        job_config = bigquery.LoadJobConfig()
+        job_config.schema = [
+            bigquery.SchemaField("metric", "STRING"),
+            bigquery.SchemaField("statistic", "STRING"),
+            bigquery.SchemaField("parameter", "FLOAT64"),
+            bigquery.SchemaField("label", "STRING"),
+            bigquery.SchemaField("ci_width", "FLOAT64"),
+            bigquery.SchemaField("point", "FLOAT64"),
+            bigquery.SchemaField("lower", "FLOAT64"),
+            bigquery.SchemaField("upper", "FLOAT64"),
+        ]
+
+        if append:
+            job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_APPEND
+        else:
+            job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_TRUNCATE
+
+        client.load_table_from_json(self.to_dict()["data"], destination_table)
+
+    def to_dict(self):
+        """Return statistic results as dict."""
+
+        return attr.asdict(self)
+
+
+@attr.s(auto_attribs=True)
+class Statistic:
+    """
+    Abstract representation of a statistic.
+
+    A statistic is a transformation that accepts a table of per-client aggregates and
+    returns a table representing a summary of the aggregates with respect to the branches
+    of the experiment.
+    """
+
+    metrics: List[str]
```

Comment:
I think it probably makes sense to have a class that represents a combination of a Metric and a Statistic since there's not an obvious reason for one to belong to the other.